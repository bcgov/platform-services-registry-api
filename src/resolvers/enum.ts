export const Cluster = {
  CLAB: "clab",
  KLAB: "klab",
  SILVER: "silver",
  GOLD: "gold",
  GOLDDR: "golddr",
  KLAB2: "klab2",
  EMERALD: "emerald"
};

export const Environment = {
  PRODUCTION: "production",
  TEST: "test",
  DEVELOPMENT: "development",
  TOOLS: "tools"
};

export const Platform = {
  PRIVATE_CLOUD: "private-cloud",
  PUBLIC_CLOUD: "public-cloud"
};

export const ProjectStatus = {
  CREATE_REQUEST: "create-request",
  ACTIVE: "active",
  INACTIVE: "inactive"
};

export const PublicCloudPlatform = {
  GOOGLE: "google-cloud",
  AWS: "amazon-web-services"
};

export const DefaultCpuOptions = {
  CPU_REQUEST_0_5_LIMIT_1_5: { cpuRequests: 0.5, cpuLimits: 1.5 },
  CPU_REQUEST_1_LIMIT_2: { cpuRequests: 1, cpuLimits: 2 },
  CPU_REQUEST_2_LIMIT_4: { cpuRequests: 2, cpuLimits: 4 },
  CPU_REQUEST_4_LIMIT_8: { cpuRequests: 4, cpuLimits: 8 },
  CPU_REQUEST_8_LIMIT_16: { cpuRequests: 8, cpuLimits: 16 },
  CPU_REQUEST_16_LIMIT_32: { cpuRequests: 16, cpuLimits: 32 },
  CPU_REQUEST_32_LIMIT_64: { cpuRequests: 32, cpuLimits: 64 }
};

export const DefaultMemoryOptions = {
  MEMORY_REQUEST_2_LIMIT_4: { memoryRequests: 2, memoryLimits: 4 },
  MEMORY_REQUEST_4_LIMIT_8: { memoryRequests: 4, memoryLimits: 8 },
  MEMORY_REQUEST_8_LIMIT_16: { memoryRequests: 8, memoryLimits: 16 },
  MEMORY_REQUEST_16_LIMIT_32: { memoryRequests: 16, memoryLimits: 32 },
  MEMORY_REQUEST_32_LIMIT_64: { memoryRequests: 32, memoryLimits: 64 },
  MEMORY_REQUEST_64_LIMIT_128: { memoryRequests: 64, memoryLimits: 128 }
};

export const DefaultStorageOptions = {
  STORAGE_1: { storageFile: 1 },
  STORAGE_2: { storageFile: 2 },
  STORAGE_4: { storageFile: 4 },
  STORAGE_16: { storageFile: 16 },
  STORAGE_32: { storageFile: 32 },
  STORAGE_64: { storageFile: 64 },
  STORAGE_124: { storageFile: 124 },
  STORAGE_256: { storageFile: 256 },
  STORAGE_512: { storageFile: 512 }
};

export const RequestDecision = {
  APPROVED: "approved",
  REJECTED: "rejected"
};

export const RequestDecisionStatus = {
  PENDING: "pending-decision",
  APPROVED: "approved",
  REJECTED: "rejected"
};

export const RequestType = {
  CREATE: "create",
  EDIT: "edit",
  DELETE: "delete"
};

export const SortOrder = {
  ASCENDING: 1,
  DESCENDING: -1
};
