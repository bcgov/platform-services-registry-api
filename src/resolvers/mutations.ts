import { MutationResolvers } from "__generated__/resolvers-types";

// Use the generated `MutationResolvers` type to type check our mutations!
const mutations: MutationResolvers = {
  Mutation: {
    // Below, we mock adding a new book. Our data set is static for this
    // example, so we won't actually modify our data.
    createUser: async (_, input, { prisma }) => {
      const { name, email, posts } = input.data;
      const postData = posts?.map((post) => {
        return { title: post.title, content: post.content || undefined }
      })
      console.log(posts)
      const result = await prisma.user.create({
        data: {
          name,
          email,
          posts: {
            create: postData
          }
        }
      });

      console.log(result)
      return result;
    }
  }
};

export default mutations;
