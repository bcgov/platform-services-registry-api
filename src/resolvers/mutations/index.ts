import {
  QueryResolvers,
  MutationResolvers,
  Resolvers,
} from '../../__generated__/resolvers-types.js';

import privateCloudProjectRequest from './privateCloud/createRequest.js';
import privateCloudRequestDecision from './privateCloud/requestDecision.js';
import privateCloudProjectEditRequest from './privateCloud/editRequest.js';
import privateCloudProjectDeleteRequest from './privateCloud/deleteRequest.js';
import privateCloudReProvisionRequest from './privateCloud/reProvisionRequest.js';
import privateCloudReProvisionProject from './privateCloud/reProvisionProject.js';

import publicCloudProjectRequest from './publicCloud/createRequest.js';
import publicCloudRequestDecision from './publicCloud/requestDecision.js';
import publicCloudProjectEditRequest from './publicCloud/editRequest.js';
import publicCloudProjectDeleteRequest from './publicCloud/deleteRequest.js';

import { createUser } from './users.js';

export default {
  Mutation: {
    privateCloudProjectRequest,
    privateCloudRequestDecision,
    privateCloudProjectEditRequest,
    privateCloudProjectDeleteRequest,
    privateCloudReProvisionRequest,
    privateCloudReProvisionProject,
    publicCloudProjectRequest,
    publicCloudRequestDecision,
    publicCloudProjectEditRequest,
    publicCloudProjectDeleteRequest,
    createUser,
  },
};
