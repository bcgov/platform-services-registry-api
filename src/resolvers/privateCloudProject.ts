import { revertCommonComponents } from '../utils/transformCommonComponents.js';

const Project = {
  projectOwner: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.projectOwnerId,
      },
    }),
  primaryTechnicalLead: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.primaryTechnicalLeadId,
      },
    }),
  secondaryTechnicalLead: async (project, _, { prisma }) =>
    project.secondaryTechnicalLeadId == null
      ? null
      : prisma.user.findUnique({
          where: {
            id: project.secondaryTechnicalLeadId,
          },
        }),
  activeEditRequest: async (project, _, { prisma }) =>
    prisma.privateCloudRequest.findFirst({
      where: {
        projectId: project.id,
        active: true,
      },
    }),
  requestHistory: async (project, _, { prisma }) =>
    prisma.privateCloudRequest.findMany({
      where: {
        projectId: project.id,
      },
    }),
  commonComponents: async (project) =>
    revertCommonComponents(project.commonComponents),
};

export default Project;
