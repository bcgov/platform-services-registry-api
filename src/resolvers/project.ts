const Project = {
  projectOwner: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.projectOwnerId
      }
    }),
  primaryTechnicalLead: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.primaryTechnicalLeadId
      }
    }),
  secondaryTechnicalLead: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.secondaryTechnicalLeadId
      }
    }),
  activeEditRequest: async (project, _, { prisma }) =>
    prisma.privateCloudRequest.findUnique({
      where: {
        projectId: project.id,
        active: true
      }
    }),
  requestHistory: async (project, _, { prisma }) =>
    prisma.privateCloudRequest.findUnique({
      where: {
        projectId: project.id
      }
    })
};

export default Project;
