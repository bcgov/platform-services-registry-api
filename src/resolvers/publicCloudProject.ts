const Project = {
  projectOwner: async (project, _, { prisma }) =>
    prisma.user.findUnique({
      where: {
        id: project.projectOwnerId,
      },
    }),
  technicalLeads: async (project, _, { prisma }) =>
    prisma.user.findMany({
      where: {
        id: {
          in: project.technicalLeadIds,
        },
      },
    }),
  activeEditRequest: async (project, _, { prisma }) =>
    prisma.publicCloudRequest.findFirst({
      where: {
        projectId: project.id,
        active: true,
      },
    }),
  requestHistory: async (project, _, { prisma }) =>
    prisma.publicCloudRequest.findMany({
      where: {
        projectId: project.id,
      },
    }),
};

export default Project;
