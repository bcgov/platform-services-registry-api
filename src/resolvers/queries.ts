import { QueryResolvers } from "__generated__/resolvers-types";

// Use the generated `QueryResolvers` type to type check our queries!
const queries: QueryResolvers = {
  Query: {
    allUsers: async (_, __, { prisma }) => {
      return await prisma.user.findMany();
    }
  }
};

export default queries;
