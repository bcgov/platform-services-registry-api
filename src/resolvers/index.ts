import {
  QueryResolvers,
  MutationResolvers,
  Resolvers
} from "../__generated__/resolvers-types";
import * as userMutations from "./mutations/users.js";
import * as requestMutations from "./mutations/requests.js";
import * as userQueries from "./queries/users.js";
import User from "./user.js";
import * as enums from "./enum.js";

const mutations: MutationResolvers = {
  Mutation: {
    ...userMutations,
    ...requestMutations
  }
};

const queries: QueryResolvers = {
  Query: { ...userQueries }
};

const resolvers: Resolvers = { ...mutations, ...queries, ...enums, User };

export default resolvers;
