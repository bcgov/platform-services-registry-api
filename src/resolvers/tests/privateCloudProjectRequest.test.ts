import "../../env";
import {
  ProjectMetaDataInput,
  Cluster,
  Ministry,
  CommonComponentsInput,
  CreateUserInput
} from "../../__generated__/resolvers-types.js";
import resolvers from "../index";
import { ApolloServer } from "@apollo/server";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { KeycloakContext, KeycloakTypeDefs } from "keycloak-connect-graphql";
import { readFileSync } from "fs";
import { DIRECTIVES } from "@graphql-codegen/typescript-mongodb";
import applyDirectiveTransformers from "../../transformers/index.js";
import { PrismaClient } from "@prisma/client";
import req from "../../auth/kauthContextMock.js";
import assert from "assert";

interface ContextValue {
  kauth: KeycloakContext;
  prisma: PrismaClient;
  authEmail: string;
  authRoles: string[];
}

const prisma = new PrismaClient();

const contextValue: ContextValue = {
  prisma: prisma,
  // @ts-ignore
  kauth: new KeycloakContext({ req }),
  authEmail: "oamar.kanji@gmail.com",
  authRoles: ["admin"]
};

const typeDefs = readFileSync("./schema.graphql", { encoding: "utf-8" });

let schema = makeExecutableSchema({
  typeDefs: [KeycloakTypeDefs, typeDefs, DIRECTIVES],
  resolvers
});

schema = applyDirectiveTransformers(schema);

const server = new ApolloServer<ContextValue>({
  schema
});

describe("Request tests", () => {
  beforeAll(async () => {});

  test("creates a private cloud project request", async () => {
    const CREATE_PRIVATE_CLOUD_PROJECT_REQUEST = `mutation PrivateCloudProjectRequest(
        $metaData: ProjectMetaDataInput!,
        $commonComponents: CommonComponentsInput!,
        $projectOwner: CreateUserInput!,
        $primaryTechnicalLead: CreateUserInput!,
        $secondaryTechnicalLead: CreateUserInput
      ) {
        privateCloudProjectRequest(
          metaData: $metaData,
          commonComponents: $commonComponents,
          projectOwner: $projectOwner,
          primaryTechnicalLead: $primaryTechnicalLead,
          secondaryTechnicalLead: $secondaryTechnicalLead
        ) {
        id
      }
    }`;

    const metaData: ProjectMetaDataInput = {
      cluster: "CLAB" as Cluster,
      description: "test description",
      ministry: "AGRI" as Ministry,
      name: "test project"
    };

    const commonComponents: CommonComponentsInput = {
      other: "test"
    };

    const projectOwner: CreateUserInput = {
      email: "oamar.kanji@gov.bc.ca",
      firstName: "Oamar",
      githubId: "okanji",
      lastName: "Kanji",
      ministry: "AGRI" as Ministry
    };

    const primaryTechnicalLead: CreateUserInput = {
      email: "xyz@test.com",
      firstName: "Jane",
      githubId: "test123",
      lastName: "Smith",
      ministry: "FIN" as Ministry
    };

    const response = await server.executeOperation(
      {
        query: CREATE_PRIVATE_CLOUD_PROJECT_REQUEST,
        variables: {
          metaData,
          commonComponents,
          projectOwner,
          primaryTechnicalLead
        }
      },
      {
        contextValue
      }
    );

    assert(response.body.kind === "single");
    expect(response.body.singleResult.errors).toBeUndefined();
    expect(response).toMatchSnapshot();
   
    // const newRequest = await prisma.privateCloudRequest.findUnique({
    //   where: {
    // });

    // 
  });
});
