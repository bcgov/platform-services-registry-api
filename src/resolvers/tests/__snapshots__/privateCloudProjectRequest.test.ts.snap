// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request tests creates a private cloud project request 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "privateCloudProjectRequest": {
          "id": "63bacabc9b4af0585352989e",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
